### PR TITLE
[Console] [bugfix] Add checks for shortcut options provided at `InputOption::__construct()`

### PR DESCRIPTION
**Pull request by phansys**
| Q | A |
| --- | --- |
| Branch | 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | kinda |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

File: src/Symfony/Component/Console/Tests/Input/InputDefinitionTest.php [hunk 1]
```diff
@@ -265,6 +265,17 @@ public function testAddDuplicateShortcutOption()
...
 
+    /**
+     * @expectedException        \LogicException
+     * @expectedExceptionMessage Invalid shortcut option "ff", it must be formed by a single char.
+     */
+    public function testAddWrongLengthShortcutOption()
+    {
+        new InputDefinition(array(
+            new InputOption('foo', 'ff', InputOption::VALUE_REQUIRED, 'The wrong length foo option'),
+        ));
```
Comment (iltar) [inline src/Symfony/Component/Console/Tests/Input/InputDefinitionTest.php]: Can you add the following test cases if you consider them a correct test?
```
ff|
f|ff
|ff
a|b|a
a|bc|de|f
```

**Comment by ogizanagi** [comment]
Actually, [Symfony does not enforce options shortcuts to be formed by a single char](https://github.com/symfony/symfony/pull/15122#issuecomment-115883976), but if it should, why not throwing an `\InvalidArgumentException` in `InputOption::__construct()` instead of `InputDefinition::addOption()` ?

Also, as pointed by @iltar, the current fix won't properly throw an exception for something like:

```php
new InputDefinition(array(
    new InputOption('foo', 'ff|f', InputOption::VALUE_REQUIRED, 'The wrong length foo option'),
));
```

**Comment by phansys** [comment]
@iltar, @ogizanagi; thank you for the feedback.
I think we must add some extra checks, like a list of allowed chars by instance.
The statement pointed by @xabbuh isn't completely true:
>we do not enforce that anywhere

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Input/ArgvInput.php#L125-L127

I know this PR can not be merged "as is", so I hope to hear more thoughts.

**Comment by phansys** [comment]
I've added some checks and tests cases. I don't know if they can be considered as bugfixes or if they must be taken as deprecations, it depends on how widely these allowed options are used, but I suspect there aren't too valid cases for them.
BTW, I'm facing issues with array sort at CI, but I can't reproduce them at 3v4l by instance: https://3v4l.org/mGbR9. I'll keep searching for a way that works the same in all the supported PHP versions.

**Comment by phansys** [comment]
Any news here? Ping @ogizanagi, @iltar; I'd like to know your thoughts.

**Comment by ogizanagi** [comment]
@phansys : Honestly, I don't understand most of those changes, nor I understand the meaning behind the term "level" in this context... 😕

- How `f|ff` can be valid, if a shortcut option `must be formed by a single char` ? Those are two available distinct shortcuts: `-f` and `-ff`, aren't they ? So at least `-ff` is invalid.
- What are the reasons behind the `Invalid shortcut option "%s", its levels must be ordered ascending` exception ? If it matters, why not simply ordering the shortcuts in the constructor ?
- Is the `Invalid shortcut option "%s", its levels must not be repeated` exception really pertinent ? Shouldn't we simply ignore duplicated shortcuts ?
- The `An option shortcut cannot be formed only with "|" chars, since they are used as level separators.` and `An option shortcut cannot be formed only with "-" chars.` are pretty obvious, and anyone trying to do something like this is already trying to do something weird. Does he deserve a such specific hint ?
- Using `array_filter` for the `An option shortcut cannot be empty` test looks flawed, because (even if it looks like a bad idea) someone may use `-0` as a shortcut. `array_filter` will return an empty array for an array of shortcuts like `[0]`.

Maybe I'm missing something about this PR, but right now I'm kind of confused :/

**Comment by phansys** [comment]
@ogizanagi, thanks for your detailed feedback. The usage of term "level" at this context can be simply explained with the argument `-v` available in all Symfony commands, see [Verbosity Levels](http://symfony.com/doc/current/console/verbosity.html).
1. `f|ff` can be considered valid by the presence of the level separator (`|`), which implicitly indicates that this is an argument which support distinct levels, so multiple single chars are allowed as a way to represent each level. By instance `f` could be considered as level 1 and `ff` as level 2.
2.  The goal behind the `Invalid shortcut option "%s", its levels must be ordered ascending` exception is to avoid unsorted leveled shortcuts, which goes against the concept of levels. By instance, `f|ff` will be considered right while `ff|f` not.
3. `Invalid shortcut option "%s", its levels must not be repeated` is very similar to the previous one, it avoids shortcuts like `f|ff|ff`. Note that the level 2 of `f` (`ff`) is repeated.
4. These blacklisted chars (`|`, `-`) are just my first thought, I think we must define and document which characters have an special meaning.
5. Sure, the current codebase supports just integers as option shortcuts, that applies to the previous concern about which characters should be supported.

I hope to kill your doubts, thank you again.

**Comment by ogizanagi** [comment]
@phansys : But verbosity levels represents different degrees of verbosity whereas `f|ff` simply represents two shortcuts for a single option. As far as I know, Symfony do not support levels for anything else than verbosity, and is ["hardcoded" in the `Application` class](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Application.php#L788-L798). 😕

**Comment by fabpot** [comment]
Verbosity is a very very specific case and cannot be replicated outside of core. So, this should not be supported in any way.

**Comment by fabpot** [comment]
Can we close this one? I'm not sure I understand what if fixes.

**Comment by phansys** [comment]
This PR tries to solve the inconsistency about the possibility to use a shortcut reference multiple times, like in the `verbose` option.

**Comment by mykiwi** [comment]
I tried your PR, but I did not succeed to understand how to play with it correctly.

If we want to be able to have an option like the verbosity, maybe we could add a new constant `InputOption::VALUE_LEVEL` and limit option's shortcut to 1 character. To play with it could be like `-v`

    bin/console command          # option's value would be 0
    bin/console command --abcd=2 # option's value would be 2
    bin/console command -aaa     # option's value would be 3

And why not like this too:

    bin/console command -a -a    # option's value would be 2

It's just an idea. I don't know if it would be easy to implement it.

**Comment by chalasr** [comment]
Closing here for the reasons explained by @ogizanagi and @fabpot.

**Pull request closed without merging**